compound documents

Archelyst opened the pull request:
This is prove of concept for an implementation of compound documents.
- I find the current `include_data` argument misleading and changed it to `include_resource_object`.
- With the `include_data` argument you can now include that data in the includes of the document.
- In order to do that you need to specify `included_schemas` in `Meta` like so:

```
class Foo(Schema):
    class Meta:
        type_ = 'foos'
        self_url = '/foos/{foo_id}/'
        self_url_kwargs = {'foo_id': '<id>'}
        included_schemas = {'bars': BarSchema()}
    bar = fields.Relationship(
        related_url='/bars/{bar_id}/',
        related_url_kwargs={'bar_id': '<bar_id>'},
        include_data=True,
        type_='bars')
```

When I then dump a foo with `include_data=True`, the included bar will be put into the included list of the document.

What's still problematic about my approach is, that the serialization of bar happens in a post_dump hook. Hence, the errors are not handled like other errors.

Any thoughts on this? Suggestions?

Archelyst committed:
prove of concept for compound documents

Archelyst committed:
moved serialization of included items into the field serialization phase

Archelyst committed:
added include_data parameter to schema instead of field

Archelyst commented:
I just improved my implementation:
1. Serialization of included data takes place during other serialization so that validation errors are handled correctly.
2. Removed `include_data` parameter from Relationship fields as from JSONAPI perspective it only makes sense to determine what to include when dumping, not at definition time.
3. Schema for included items is now declared on the field, so that everything is in one place.

Example:

```
   my_field = fields.Relationship(
        related_url='/foo/{foo_id}/',
        related_url_kwargs={'foo_id': '<foo.id>'},
        type_='foos',
        schema='FooSchema')
```

Please tell me if you can imagine merging this, than I'd fix and add tests.

sloria commented:
Thanks for the PR! I would certainly merge this feature once it's fixed up with docs and tests.

sloria commented on marshmallow_jsonapi/fields.py:
```diff
@@ -22,6 +25,12 @@ class BaseRelationship(Field):
     pass
 
 
+def stringify(value):
```
Would be good to "privatize" this by renaming to `_stringify`.

Archelyst replied:
Ok. Why do we actually need this? I realized that while the IDs in the main data object are integers, the IDs in the linkages are strings.

sloria replied:
According to the spec:

>  The values of the id and type members MUST be strings.

http://jsonapi.org/format/#document-resource-object-identification

We _should_ be stringifying IDs in the main data object as well. =/

sloria commented on marshmallow_jsonapi/fields.py:
```diff
@@ -49,8 +58,9 @@ class Relationship(BaseRelationship):
     :param dict self_url_kwargs: Replacement fields for `self_url`. String arguments
         enclosed in `< >` will be interpreted as attributes to pull from the target object.
-    :param bool include_data: Whether to include a resource linkage
+    :param bool include_resource_object: Whether to include a resource linkage
```
Would `include_resource_linkage` be more descriptive?

Archelyst replied:
Sounds good for me.

Archelyst committed:
renamed include_resource_object to include_resource_linkage

Archelyst committed:
added tests for compound documents

Archelyst committed:
added documentation for compound documents

Archelyst committed:
fixed compatibility with 2.7

Archelyst committed:
privatized stringify()

Archelyst commented:
Ok, that should be it. Any more things to consider?

sloria commented on docs/quickstart.rst:
```diff
@@ -140,6 +140,69 @@ You can serialize `resource linkages <http://jsonapi.org/format/#document-resour
+    #     }
+    #     "included": [
+    #         {
+    #             "data": {
```
From http://jsonapi.org/format/#document-compound-documents , it appears that included data should not have the `data` envelope.

Archelyst replied:
Good catch, will fix that.

Archelyst replied:
Done.

Archelyst committed:
removed 'data' envelope from included resources

sloria commented on marshmallow_jsonapi/schema.py:
```diff
@@ -76,7 +76,22 @@ class Meta:
+                    raise ValueError('A schema is required to serialize "%s"' % field_name)
+                field.include_data = True
+            else:
+                field.include_data = False
```
This will raise an error for fields that define `__slots__` (which is proposed here: https://github.com/marshmallow-code/marshmallow/pull/422). 

Can we safely remove this `else` block?

Archelyst replied:
I don't think so, because unfortunately fields are only instantiated once, while the schema can be instantiated multiple times with different `include_data` arguments, so we need to have a way to set `include_data` of fields to `False`. I'll change it to

```
elif isinstance(field, BaseRelationship):
    field.include_data = False
```

That should prevent the mentioned error.

sloria commented on marshmallow_jsonapi/schema.py:
```diff
@@ -76,7 +76,22 @@ class Meta:
+        for field_name, field in self.fields.items():
+            if field_name in self.include_data:
+                if not isinstance(field, BaseRelationship):
+                    raise ValueError('Can only include relationships. "%s" is a "%s"'
```
Minor: for consistency with the rest of the codebase, use `str.format` rather than `%` for string formatting.

sloria commented on tests/test_schema.py:
```diff
@@ -106,6 +121,35 @@ def test_dump_to(self, post):
+        data = PostSchema(include_data=()).dump(post).data
+        assert 'included' not in data
+
+
```
Can you also test cases where validation fails (for both `many=True` and `many=False`)?

Archelyst replied:
Could you give me an example of a scenario where a validation fails? I'm having a hard time trying to provoke a ValidationError when dumping (and this compound documents are dumped only).

sloria commented on marshmallow_jsonapi/fields.py:
```diff
@@ -156,9 +176,25 @@ def _serialize(self, value, attr, obj):
+            if self.many:
+                for item in value:
+                    # items need to be dumped separately,
+                    # othwerwise we'd get a JSONAPI list
```
Would the following code work?

``` python
                result = self.schema.dump(value, many=True)
                if result.errors:
                    raise ValidationError(result.errors)
                self.root.included_data.extend(result.data['data'])
```

Archelyst replied:
I think it would. You find that clearer? Or maybe that's faster? It's somehow odd to me to first make a jsonapi list just to unpack it afterwards.

Archelyst replied:
In my last commit I changed this block. Are you ok with that now?

sloria commented:
Pass finished. This is looking good! Just a few more things to address above.

ghost commented:
just a little suggestion since this hasn't been merged yet, that it would be very nice additional feature to this if we can pass arguments for the passed `schema` so we can do further configuration, e.g.

```
my_field = fields.Relationship(
        related_url='/foo/{foo_id}/',
        related_url_kwargs={'foo_id': '<foo.id>'},
        type_='foos',
        schema='FooSchema',
        schema_args={'only': ('field1', 'field2')})
```

then somewhere inside, use `**schema_args` as `FooSchema` gets invoked.

I was originally looking onto how to implement [Sparse Fieldsets](http://jsonapi.org/format/#fetching-sparse-fieldsets) and apparently you can do it as you dump the schema with the `only` parameter, but you wont be able to do the same for the `included` resources but by passing arguments to the relationship schema, you may be able to.

Archelyst commented:
There is an open PR for marshmallow (https://github.com/marshmallow-code/marshmallow/pull/405 https://github.com/marshmallow-code/marshmallow/pull/405) that would allow to configure the only parameter for nested schemas. As soon as that is merged, I would imagine, we can extend it to marshmallow-jsonapi. Please have a look at it and support it.

> Am 17.05.2016 um 09:18 schrieb Charles Jordan Cabero notifications@github.com:
> 
> just a little suggestion since this hasn't been merged yet, that it would be very nice additional feature to this if we can pass arguments for the passed schema so we can do further configuration, e.g.
> 
> my_field = fields.Relationship(
>         related_url='/foo/{foo_id}/',
>         related_url_kwargs={'foo_id': '<foo.id>'},
>         type_='foos',
>         schema='FooSchema',
>         schema_args={'only': ('field1', 'field2')})
> then somewhere inside, use **schema_args as FooSchema gets invoked.
> 
> I was originally looking onto how to implement Sparse Fieldsets http://jsonapi.org/format/#fetching-sparse-fieldsets and apparently you can do it as you dump the schema with the only parameter, but you wont be able to do the same for the included resources but by passing arguments to the relationship schema, you may be able to.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub https://github.com/marshmallow-code/marshmallow-jsonapi/pull/40#issuecomment-219638416

Archelyst committed:
self-referential relationships

Archelyst committed:
made inluded objects in compound documents unique

v-ken commented:
Just noticed a small error in the [documentation](https://github.com/Tim-Erwin/marshmallow-jsonapi/blob/compound-documents/docs/quickstart.rst)

`ArticleSchema(include=('comments',)).dump(article).data`

should be

`ArticleSchema(include_data=('comments',)).dump(article).data`

woodb commented:
@chavz Great idea! I think it might be worth capturing that as a separate issue so that we don't have to wait too long for this PR to get merged in.

Is there anything that I can do to help move this along? What is left to do?

woodb committed:
Fixed minor issue in compound documents example

Per @v-ken's comment, this commit fixes a small error in the documentation of
the compound documents (i.e. `included`) feature in
PR marshmallow-code/marshmallow-jsonapi#40

https://github.com/marshmallow-code/marshmallow-jsonapi/pull/40#issuecomment-225129348

woodb commented:
@Tim-Erwin I've submitted a PR to your branch to fix the documentation issue @v-ken mentioned

woodb committed:
Added pyenv .python-version file to gitignore

woodb committed:
Additional tests for recursive many includes

woodb committed:
Added support for many for "self" compound docs

woodb commented:
@Tim-Erwin I have added support for `many=True` when the `schema="self"` in the PR above to this branch (this includes the doc fixes, so I've closed that one out)

woodb committed:
Deeply nested relationships now also can be included

woodb commented:
Just curious how we might be able to keep pushing this PR along and get it merged in 👍 

Any thoughts?

Archelyst committed:
Merge remote-tracking branch 'woodb/compound-documents-many-fix' into compound-documents

Archelyst commented:
@v-ken , thank for the hint. @woodb , your PR is in, thanks a lot. @sloria , do you still see any issues with this? I'm using it in production successfully already. The only thing missing as far as I'm concerned is a line for @woodb in the authors' file :)

sloria commented:
Thank you both for your hard work on this!

Archelyst commented:
Merged, awesome 🎉  thanks

neaket360pi commented on marshmallow_jsonapi/fields.py:
```diff
@@ -63,20 +76,42 @@ def __init__(
+        if isinstance(self.__schema, SchemaABC):
+            return self.__schema
+        if isinstance(self.__schema, type) and issubclass(self.__schema, SchemaABC):
+            self.__schema = self.__schema(many=self.many)
```
@Tim-Erwin Have you tested with `self.many=True`?   I think this line should be changed to have `many=False`.
```python
self.__schema = self.__schema()
```
The code currently iterates over the relationship's data on [line 196](https://github.com/marshmallow-code/marshmallow-jsonapi/pull/40/files#diff-b882a3f5a63ddce7dd43a950cd1d20d1R198) and dumps each item separately.

CaptObvious replied:
Is this why it breaks when trying to include data that is ```many=False``` when the root object is ```many=True```?

jordal commented:
@Tim-Erwin I've also run into a use case for passing ``only`` args to the Schema used to serialize the relationship data. I saw that https://github.com/marshmallow-code/marshmallow/pull/468 which succeeds #405 was merged. What's required to integrate this with relationship fields in compound documents?

> There is an open PR for marshmallow (marshmallow-code/marshmallow#405 marshmallow-code/marshmallow#405) that would allow to configure the only parameter for nested schemas. As soon as that is merged, I would imagine, we can extend it to marshmallow-jsonapi. Please have a look at it and support it.
> 
> > Am 17.05.2016 um 09:18 schrieb Charles Jordan Cabero notifications@github.com:
> > 
> > just a little suggestion since this hasn't been merged yet, that it would be very nice additional feature to this if we can pass arguments for the passed schema so we can do further configuration, e.g.
> > 
> > my_field = fields.Relationship(
> > related_url='/foo/{foo_id}/',
> > related_url_kwargs={'foo_id': '<foo.id>'},
> > type_='foos',
> > schema='FooSchema',
> > schema_args={'only': ('field1', 'field2')})
> > then somewhere inside, use **schema_args as FooSchema gets invoked.
> > 
> > I was originally looking onto how to implement Sparse Fieldsets http://jsonapi.org/format/#fetching-sparse-fieldsets and apparently you can do it as you dump the schema with the only parameter, but you wont be able to do the same for the included resources but by passing arguments to the relationship schema, you may be able to.
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly or view it on GitHub #40 (comment)

BillBrower commented:
@jordal I'm pretty sure you can already pass Schema arguments like this:

```python
my_field = fields.Relationship(
    related_url='/foo/{foo_id}/',
    related_url_kwargs={'foo_id': '<foo.id>'},
    type_='foos',
    schema=FooSchema(only=('field1', 'field2')
)
```
